### PR TITLE
Added auth package and implemented basic auth

### PR DESCRIPTION
**Pull request by dusnm**
Signed-off-by: Dušan Mitrović <dmitrovic@nanointeractive.com>

This feature adds basic authorization capabilities.

### CODE REVIEW

**Review by CerealKiller97 — APPROVED** [review]

File: auth/token.go [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+	components := strings.SplitN(header, " ", 2)
+
+	if len(components) < 2 {
+		return nil, fmt.Errorf("Invalid header")
+	}
+
+	authType := components[0]
+	token := components[1]
+
+	if !strings.EqualFold(authType, AuthTypeBearer) {
```
Comment (CodeLieutenant) [inline auth/token.go]: Make AuthTypeBearer custom type

```
type AuthType string

const (
   AuthTypeBearer AuthType = "bearer"
)
```

File: auth/token.go [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+	"strings"
+
+	"github.com/nano-interactive/go-utils"
+)
+
+const (
+	AuthTypeBearer = "bearer"
+)
+
+func ParseTokenFromHeader(header string) ([]byte, error) {
```
Comment (CodeLieutenant) [inline auth/token.go]: Change function signature

```
func ParseTokenFromHeader(header string, authType AuthType) ([]byte, error)
```

File: auth/token.go [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+
+	var tokenBytes [32]byte
+
+	if base64.RawURLEncoding.DecodedLen(len(token)) != 32 {
+		return nil, fmt.Errorf("Invalid token")
+	}
+
+	_, err := base64.RawURLEncoding.Decode(tokenBytes[:], utils.UnsafeBytes(token))
+	if err != nil {
+		return nil, fmt.Errorf("Invalid token")
```
Comment (CodeLieutenant) [inline auth/token.go]: Remove `fmt.Errorf` it causes unnecessary allocation.

```
var ErrInvalidToken = errors.New("invalid token")
```

File: auth/token.go [hunk 4]
```diff
@@ -0,0 +1,51 @@
...
+	}
+
+	return tokenBytes[:], nil
+}
+
+func CompareTokens(token1 []byte, token2 []byte) int {
+	// Now, you may wonder why we need this function if all ti does
```
Comment (CodeLieutenant) [inline auth/token.go]: Remove this comment

File: auth/token.go [hunk 5]
```diff
@@ -0,0 +1,51 @@
...
+
+	"github.com/nano-interactive/go-utils"
+)
+
+const (
+	AuthTypeBearer = "bearer"
+)
+
+func ParseTokenFromHeader(header string) ([]byte, error) {
+	components := strings.SplitN(header, " ", 2)
```
Comment (CodeLieutenant) [inline auth/token.go:13]: Try to optimize this allocation if you can

**Review by CodeLieutenant — CHANGES_REQUESTED** [review]

File: auth/token.go (hunk 5 above)
Comment (dusnm) [reply auth/token.go:13]: WONTFIX: Allocations are inevitable, code comprehensibility is not. I sincerely doubt the performance benefit outweighs clarity in this case.

**Review by CodeLieutenant — APPROVED** [review]

**Pull request merged**